librz/core/cprint.c

```diff
@@ -156,4 +156,240 @@ RZ_API RZ_OWN char *rz_core_esil_of_hex(RzCore *core, ut8 *hex, int len) {
+ * \brief Print dump at \p addr
+ * \param n Word size by bytes (1,2,4,8)
+ * \param len Dump bytes length
+ * \param format Print format, such as RZ_CORE_PRINT_FORMAT_TYPE_HEXADECIMAL
+ */
+RZ_API char *rz_core_print_dump_str(RZ_NONNULL RzCore *core, RZ_NULLABLE RzCmdStateOutput *state,
+	ut64 addr, ut8 n, int len, RzCorePrintFormatType format) {
+	rz_return_val_if_fail(core, false);
+	if (!len) {
+		return NULL;
+	}
+	st8 base = format_type_to_base(format, n);
+	if (!base) {
+		return NULL;
+	}
+	len_fixup(core, &addr, &len);
+	ut8 *buffer = malloc(len);
+	if (!buffer) {
+		return NULL;
+	}
+
+	char *pstr = NULL;
+	rz_io_read_at(core->io, addr, buffer, len);
+	rz_print_init_rowoffsets(core->print);
+	core->print->use_comments = false;
+	RzOutputMode mode = state ? state->mode : RZ_OUTPUT_MODE_STANDARD;
+
+	switch (mode) {
+	case RZ_OUTPUT_MODE_JSON:
+		pstr = rz_print_jsondump_str(core->print, buffer, len, n * 8);
+		break;
+	case RZ_OUTPUT_MODE_STANDARD:
+		fix_size_from_format(format, &n);
+		pstr = rz_print_hexdump_str(core->print, addr,
+			buffer, len, base, (int)n, 1);
+		break;
+	default:
+		rz_warn_if_reached();
+		free(buffer);
+		return NULL;
+	}
+	free(buffer);
+	return pstr;
+}
+
+/**
+ * \brief Print hexdump at \p addr, but maybe print hexdiff if (diff.from or diff.to), \see "el diff"
+ * \param len Dump bytes length
+ */
+RZ_API char *rz_core_print_hexdump_or_hexdiff_str(RZ_NONNULL RzCore *core, RZ_NULLABLE RzCmdStateOutput *state, ut64 addr, int len) {
+	rz_return_val_if_fail(core, false);
+	if (!len) {
+		return NULL;
+	}
+
+	RzOutputMode mode = state ? state->mode : RZ_OUTPUT_MODE_STANDARD;
+	char *pstr = NULL;
+	switch (mode) {
+	case RZ_OUTPUT_MODE_STANDARD: {
+		ut64 from = rz_config_get_i(core->config, "diff.from");
+		ut64 to = rz_config_get_i(core->config, "diff.to");
+		if (from == to && !from) {
+			len_fixup(core, &addr, &len);
+			ut8 *buffer = malloc(len);
+			if (!buffer) {
+				return NULL;
+			}
+			rz_io_read_at(core->io, addr, buffer, len);
+			pstr = rz_print_hexdump_str(core->print, rz_core_pava(core, addr), buffer, len, 16, 1, 1);
+			free(buffer);
+		} else {
+			pstr = rz_core_print_hexdump_diff_str(core, addr, addr + to - from, len);
+		}
+		break;
+	}
+	case RZ_OUTPUT_MODE_JSON:
+		pstr = rz_print_jsondump_str(core->print, core->block, len, 8);
+		break;
+	default:
+		rz_warn_if_reached();
+		return NULL;
+	}
+	return pstr;
+}
+
+static inline char *ut64_to_hex(const ut64 x, const ut8 width) {
+	RzStrBuf *sb = rz_strbuf_new(NULL);
+	rz_strbuf_appendf(sb, "%" PFMT64x, x);
+	ut8 len = rz_strbuf_length(sb);
+	if (len < width) {
+		rz_strbuf_prepend(sb, rz_str_pad('0', width - len));
+	}
+	rz_strbuf_prepend(sb, "0x");
+	return rz_strbuf_drain(sb);
+}
+
+/**
+ * \brief Hexdump at \p addr
+ * \param len Dump bytes length
+ * \param size Word size by bytes (1,2,4,8)
+ * \return Hexdump string
+ */
+RZ_API RZ_OWN char *rz_core_print_hexdump_byline_str(RZ_NONNULL RzCore *core, RZ_NULLABLE RzCmdStateOutput *state,
```

Comment:
Since these functions aren't printing anymore, maybe we could give them slightly different name? Any suggestions? @ret2libc @wargio @thestr4ng3r @wingdeans